services/airflow/src/dags/datenspende/add_vaccination_data/add_vaccination_data.py

```diff
@@ -0,0 +1,480 @@
+import pandas as pd
+import ramda as R
+import numpy as np
+from datetime import datetime
+from psycopg2.sql import SQL
+from src.lib.dag_helpers import (
+    execute_query_and_return_dataframe,
+    connect_to_db_and_upsert_pandas_dataframe_on_constraint,
+)
+from src.lib.test_helpers import set_env_variable_from_dag_config_if_present
+from postgres_helpers import create_db_context, teardown_db_context, DBContext
+
+
+def add_vaccination_data_to_homogenized_feature_table(**kwargs):
+    R.pipe(
+        set_env_variable_from_dag_config_if_present("TARGET_DB"),
+        load_and_transform_vaccination_data_for_each_user,
+        connect_to_db_and_upsert_pandas_dataframe_on_constraint(
+            "datenspende_derivatives", "homogenized_features", ["unique_answers"]
+        ),
+    )(kwargs)
+
+
+def load_and_transform_vaccination_data_for_each_user(*_):
+    db_context = create_db_context()
+    vaccination_data = load_vaccinations(db_context)
+    feature_data = load_feature_data(db_context)
+    teardown_db_context(db_context)
+
+    return R.pipe(
+        R.invoker(0, "itertuples"),
+        R.reduce(vaccination_data_reducer(vaccination_data), []),
+        pd.DataFrame,
+    )(feature_data)
+
+
+@R.curry
+def vaccination_data_reducer(vaccination_data, res, feature_data_row):
+    try:
+        user_vaccination_data = vaccination_data.set_index("user_id").loc[
+            [feature_data_row.user_id]
+        ]
+    except KeyError:
+        return res + [
+            {
+                "user_id": feature_data_row.user_id,
+                "test_week_start": feature_data_row.test_week_start,
+                "administered_vaccine_doses": None,
+                "days_since_last_dose": None,
+            }
+        ]
+
+    dose_administration_dates = user_vaccination_data[
+        ["first_dose", "second_dose", "third_dose"]
+    ].values[0]
+
+    past_dose_administration_dates = [
+        pd.Timestamp(date)
+        for date in dose_administration_dates
+        if pd.Timestamp(date) < pd.Timestamp(feature_data_row.test_week_start)
+    ]
+
+    number_of_administered_doses = len(past_dose_administration_dates)
+    try:
+        days_since_last_dose = (
+            pd.Timestamp(feature_data_row.test_week_start)
+            - max(past_dose_administration_dates)
+        ).days
```

Comment:
This is where numer of administered doses and days since last vaccination are calculated.

Reply:
okay. so this means that you compute some form of maximum possible `days_since_last_dose` since `max(past_dose_administration_dates)` gives you the first day of the month of the last vaccination dose. but the person could have also gotten it on the last day of the months. its okay to do it this way i think, just sth. to keep in mind when writing the paper 🙂